src/Renderer.js

```diff
@@ -344,114 +403,52 @@ export default class Renderer {
-          Matrix.rotate(m, m, spr.scratchToRad(spr.direction));
-          break;
-        }
-        case Sprite.RotationStyle.LEFT_RIGHT: {
-          if (spr.direction < 0) Matrix.scale(m, m, -1, 1);
-          break;
-        }
-      }
-
-      const spriteScale = spr.size / 100;
-      Matrix.scale(m, m, spriteScale, spriteScale);
-    }
-
-    const scalingFactor = 1 / spr.costume.resolution;
-    // Rotation centers are in non-Scratch space (positive y-values = down),
-    // but these transforms are in Scratch space (negative y-values = down).
-    Matrix.translate(
-      m,
-      m,
-      -spr.costume.center.x * scalingFactor,
-      (spr.costume.center.y - spr.costume.height) * scalingFactor
-    );
-    Matrix.scale(
-      m,
-      m,
-      spr.costume.width * scalingFactor,
-      spr.costume.height * scalingFactor
-    );
-
-    return m;
-  }
-
-  // Calculate the transform matrix for a speech bubble attached to a sprite.
-  _calculateSpeechBubbleMatrix(spr, speechBubbleSkin) {
-    const sprBounds = this.getBoundingBox(spr);
-    let x;
-    if (
-      speechBubbleSkin.width + sprBounds.right >
-      this.project.stage.width / 2
-    ) {
-      x = sprBounds.left - speechBubbleSkin.width;
-      speechBubbleSkin.flipped = true;
-    } else {
-      x = sprBounds.right;
-      speechBubbleSkin.flipped = false;
-    }
-    x = Math.round(x - speechBubbleSkin.offsetX);
-    const y = Math.round(sprBounds.top - speechBubbleSkin.offsetY);
+    // Enable color masking mode if set
+    if (Array.isArray(colorMask))
+      this.gl.uniform4fv(this._currentShader.uniforms.u_colorMask, colorMask);
 
-    const m = Matrix.create();
-    Matrix.translate(m, m, x, y);
-    Matrix.scale(m, m, speechBubbleSkin.width, speechBubbleSkin.height);
-
-    return m;
+    // Actually draw the skin
+    this.gl.drawArrays(this.gl.TRIANGLES, 0, 6);
   }
 
   renderSprite(sprite, options) {
     const spriteScale = Object.prototype.hasOwnProperty.call(sprite, "size")
       ? sprite.size / 100
       : 1;
 
-    this._setSkinUniforms(
-      this._skinCache.getSkin(sprite.costume),
+    this._renderSkin(
+      this._getSkin(sprite.costume),
       options.drawMode,
-      this._calculateSpriteMatrix(sprite),
+      this._getDrawable(sprite).getMatrix(),
       spriteScale,
       sprite.effects,
-      options.effectMask
+      options.effectMask,
+      options.colorMask
     );
-    if (Array.isArray(options.colorMask))
-      this.gl.uniform4fv(
-        this._currentShader.uniforms.u_colorMask,
-        options.colorMask
-      );
-    this.gl.drawArrays(this.gl.TRIANGLES, 0, 6);
 
     if (
       options.renderSpeechBubbles &&
       sprite._speechBubble &&
       sprite._speechBubble.text !== ""
     ) {
-      const speechBubbleSkin = this._skinCache.getSkin(sprite._speechBubble);
+      const speechBubbleSkin = this._getSkin(sprite._speechBubble);
 
-      this._setSkinUniforms(
+      this.renderSkin(
```

Comment:
```suggestion
      this._renderSkin(
```

Reply:
Incorrect method name. Errors when using `this.touching("edge")` while a speech bubble is active.

(I got very lucky! My test project happened to include this case. We could consider using Typescript to avoid these kinds of errors and also to provide type hints for people using Leopard.)

Reply:
Good catch! I've opened a new issue branching off this to help avoid similar issues: #131